draft-aanchal-time-implementation-guidance.md

```diff
@@ -0,0 +1,358 @@
+    and have no means to process absolute time until synchronization with
+    these sources has completed.
+
+  * Relying on Internet timing protocols opens up the system time to
+    attack.  Recent papers show vulnerabilities in NTP [@ANTP][@ANABM][@SECNTP]
+    and SNTP [@BPHSTS] that allow attackers to maliciously alter system's world
+    clock -- pushing it into the past or even into the future.  Moreover,
+    many of these time-shifting attacks can be performed by off-path
+    attackers, who do not occupy a privileged position on the network
+    between the victim system and its time sources on the Internet.
+    Researchers have also demonstrated off-path denial of service attacks
+    on timing protocols that prevent systems from synchronizing their
+    clocks.
+
+In other words, the process of obtaining the offset necessary to
+provide a world clock creates dependencies that can be exploited.
+
+# Implementation Approaches
+
+Because absolute time relies on a shared interpretation of a value expressing
+time, the world clock is necessary when processing such values.
+
+For relative time, however, where only the rate of passage of time needs to be
+close enough to that of the other systems involved, there is no need to rely on
+the world clock when determining whether an interval has passed.
+
+Instead, by obtaining a value from the native clock when the interval has
+started only the native clock is necessary to determine when this interval
+ends. As the native clock does not rely on any external time sources, the
+implementation becomes resistant to the difficulties of coordinating with these
+sources.
+
+However, using the native clock in this way comes with a caveat. Since the
+native clock is not subject to any adjustments by timing protocols, it is not
+adjusted for the error introduced by clock drift. While this is likely of
+little consequence for short intervals, it may become significant for intervals
+that span long periods of time.
+
+The choice of clock to be used is situation-specific. If a certain amount of
+clock drift can be tolerated or if time intervals are short, implementors may
+prefer to use the native clock. However, if precise timing over long periods is
+required, then the implementors have no choice but to fall back to world clock
+
+# Accessing the Native Clock on Selected Operating Systems
+
+In most operating systems, the standard functions to access time use the world
+clock since that is normally what users would expect. This section provides an
+overview how the native clock can be accesses on some common operating systems.
+
+## POSIX
+
+POSIX defines a system C API function which may provide native time:
+`clock_gettime()`, when used with a `clock_id` of `CLOCK_MONOTONIC`.
+
+Note that on some systems `CLOCK_MONOTONIC` is still influenced by an external
+time source (for syntonizing the clock rate) and `CLOCK_MONITONIC_RAW` needs to
```

Comment:
If I remember correctly, `CLOCK_MONOTONIC_RAW` is non-POSIX? If so, this should probably be mentioned (‘and the non-standard `CLOCK_MONOTONIC_RAW` needs to be used`)

Reply:
Agree, will do